File: packages/stripe_terminal_android/android/src/main/kotlin/com/stripeterminalreactnative/TokenProvider.kt
package com.stripeterminalreactnative

import com.facebook.react.bridge.ReactApplicationContext
import com.stripe.stripeterminal.external.callable.ConnectionTokenCallback
import com.stripe.stripeterminal.external.callable.ConnectionTokenProvider
import com.stripe.stripeterminal.external.models.ConnectionTokenException
import com.stripeterminalreactnative.ReactExtensions.sendEvent
import com.stripeterminalreactnative.ReactNativeConstants.FETCH_TOKEN_PROVIDER

class TokenProvider(private val context: ReactApplicationContext) : ConnectionTokenProvider {
    private var connectionTokenCallback: ConnectionTokenCallback? = null

    fun setConnectionToken(token: String?, error: String?) {
        try {
            if (!token.isNullOrEmpty()) {
                connectionTokenCallback?.onSuccess(token)
                connectionTokenCallback = null
            } else {
                connectionTokenCallback?.onFailure(
                    ConnectionTokenException(error ?: "", null)
                )
            }
        } catch (e: Throwable) {
            connectionTokenCallback?.onFailure(
                ConnectionTokenException("Failed to fetch connection token", e)
            )
        }
    }

    override fun fetchConnectionToken(callback: ConnectionTokenCallback) {
        connectionTokenCallback = callback
        context.sendEvent(FETCH_TOKEN_PROVIDER.listenerName, null)
    }
}

File: packages/stripe_terminal_ios/ios/Classes/Stripe Sdk/TokenProvider.swift
import StripeTerminal

enum TokenError: Error {
    case runtimeError(String)
}

class TokenProvider: ConnectionTokenProvider {
    static let shared = TokenProvider()
    var completionCallback: ConnectionTokenCompletionBlock? = nil

    static var delegate: RCTEventEmitter? = nil

    func setConnectionToken(token: String?, error: String?) {
        let unwrappedToken = token ?? ""
        if (!unwrappedToken.isEmpty) {
            self.completionCallback?(token, nil)
        } else {
            self.completionCallback?(nil, TokenError.runtimeError(error ?? "") )
        }
    }

    func fetchConnectionToken(_ completion: @escaping ConnectionTokenCompletionBlock) {
        self.completionCallback = completion
        TokenProvider.delegate?.sendEvent(withName: ReactNativeConstants.FETCH_TOKEN_PROVIDER.rawValue, body: [:])
    }
}
